core/src/main/java/org/sonarsource/sonarlint/core/AutoBinding.java

```diff
@@ -0,0 +1,408 @@
+    if (AUTOSCAN_CONFIG_FILENAME.equals(filename)) {
+      return new SonarCloudBindingClue(scannerProps.projectKey, scannerProps.organization);
+    }
+    if (scannerProps.organization != null) {
+      return new SonarCloudBindingClue(scannerProps.projectKey, scannerProps.organization);
+    }
+    if (scannerProps.serverUrl != null) {
+      if (removeEnd(scannerProps.serverUrl, "/").equals(SONARCLOUD_URL)) {
+        return new SonarCloudBindingClue(scannerProps.projectKey, null);
+      } else {
+        return new SonarQubeBindingClue(scannerProps.projectKey, scannerProps.serverUrl);
+      }
+    }
+    if (scannerProps.projectKey != null) {
+      return new UnknownBindingClue(scannerProps.projectKey);
+    }
+    return null;
+  }
+
+  private interface BindingClue {
+
+    @CheckForNull
+    String getSonarProjectKey();
+
+  }
+  private static class UnknownBindingClue implements BindingClue {
+    private final String sonarProjectKey;
+
+    private UnknownBindingClue(String sonarProjectKey) {
+      this.sonarProjectKey = sonarProjectKey;
+    }
+
+    @Override
+    public String getSonarProjectKey() {
+      return sonarProjectKey;
+    }
+  }
+
+  private static class SonarQubeBindingClue implements BindingClue {
+
+    private final String sonarProjectKey;
+    private final String serverUrl;
+
+    private SonarQubeBindingClue(@Nullable String sonarProjectKey, String serverUrl) {
+      this.sonarProjectKey = sonarProjectKey;
+      this.serverUrl = serverUrl;
+    }
+
+    @Override
+    public String getSonarProjectKey() {
+      return sonarProjectKey;
+    }
+  }
+
+  private static class SonarCloudBindingClue implements BindingClue {
+
+    private final String sonarProjectKey;
+    private final String organization;
+
+    private SonarCloudBindingClue(@Nullable String sonarProjectKey, @Nullable String organization) {
+      this.sonarProjectKey = sonarProjectKey;
+      this.organization = organization;
+    }
+
+    @Override
+    public String getSonarProjectKey() {
+      return sonarProjectKey;
+    }
+  }
+
+  private boolean checkIfValidCandidateForAutoBinding(String configScopeId) {
+    var bindingConfiguration = configReferential.getBindingConfiguration(configScopeId);
+    if (bindingConfiguration == null) {
+      // Race condition
+      LOG.debug("Configuration scope '{}' is gone. Skipping auto-binding", configScopeId);
+      return false;
+    }
+    if (isBound(bindingConfiguration)) {
+      LOG.debug("Configuration scope '{}' is already bound. Skipping.", configScopeId);
+      return false;
+    }
+    if (!bindingConfiguration.isAutoBindEnabled()) {
+      LOG.debug("Configuration scope '{}' has auto-bind disabled. Skipping.", configScopeId);
+      return false;
+    }
+    return true;
+  }
+
+  private boolean isBound(BindingConfiguration bindingConfiguration) {
+    return bindingConfiguration.getConnectionId() != null
+      && bindingConfiguration.getSonarProjectKey() != null
```

Comment:
Same remark goes here, this `isBound` could be moved to a `Binding` class (not the DTO)

Reply:
I will only move the part checking for connectionId and projectKey being non null. To check if the connection actually exists, I would need to introduce a coupling between `BindingConfiguration` and `ConnectionConfig`. As discussed, I would like to introduce at some point a concept of "LiveConnection" and "LiveBinding", that we could use in this case. That will come later.